docs(fusion): add page State Management With MobX (#176)
vaadin/docs: articles/fusion/application/state-management.asciidoc

---
title: State Management with MobX
order: 80
layout: page
---

= State Management With MobX

Application state can be managed in many different ways.
In simple applications you may not need an explicit state management strategy.
But when your application grows to have a lot of views or complex UI
where the same data needs to be displayed in multiple places
or when there are many dependencies between UI controls
it may become difficult to keep the application in a consistent state
when the user interacts with the application.
If you don't have a good state management strategy
it can be easy to introduce new bugs when changing application logic.

This is where a state management library like link:https://mobx.js.org/[MobX] can help
and make the application easier to maintain and extend.

> Here we concentrate on an approach to store all shared front-end state in a central MobX state store

Reviewer: ```suggestion
Here we concentrate on an approach to store all shared frontend state in a central MobX state store
```

Author: Done.

which will be the single source of truth for the application state.
Then it's easy to display the state data (or anything that can be derived from it) anywhere in the UI so that the data is automatically updated everywhere where it is displayed whenever the data is updated in the state store.

== Using MobX in Your Application

> Vaadin recommends to use the link:https://mobx.js.org/[MobX] library to manage front-end state in your apps.

Reviewer: ```suggestion
Vaadin recommends to use the link:https://mobx.js.org/[MobX] library to manage frontend state in your apps.
```

Author: Done.

When using LitElement that means you should extend the `MobxLitElement` class instead of `LitElement`.
When your views are based on `MobxLitElement`,
any MobX observables used in the `render()` method are automatically tracked
so that any change to those observables is automatically reflected in the view.

Creating a new Vaadin Fusion project with

> link:https://start.vaadin.com/[start.vaadin.com] (or link:https://vaadin.com/labs/cli[Vaadin CLI])

Reviewer: ```suggestion
link:https://start.vaadin.com/[Start Vaadin] (or link:https://vaadin.com/labs/cli[Vaadin CLI])
```

Author: Done.

already takes care of setting up the basics for you

> by providing a MobX store (named `AppState`) in addition to `View` and `Layout` helper base classes for your application views and layouts.

Reviewer: ```suggestion
by providing a MobX store (named `AppState`) and `View` and `Layout` helper base classes for your application views and layouts.
```

Author: Done.

Both `View` and `Layout` extend `MobxLitElement` and you may expand the `AppState` store with your own observables and actions to be used in your views.

.MobX decorators with TypeScript
[CAUTION]
====

> MobX has experimental support for decorators (`@observable`, `@action`) but using them requires a special setting `"useDefineForClassFields": true` in `tsconfig.json`. Unfortunately having that setting link:https://github.com/Polymer/lit-element/issues/855[breaks LitElement decorators] (`@property`, `@query`) so you can't use both LitElement decorators and MobX decorators in the same app. We recommend not to use MobX decorators for now in projects that use LitElement.

Reviewer: ```suggestion
MobX has experimental support for decorators (`@observable`, `@action`) but using them requires a special setting `"useDefineForClassFields": true` in `tsconfig.json`. Using that setting link:https://github.com/Polymer/lit-element/issues/855[breaks LitElement decorators] (`@property`, `@query`) so you can't use both LitElement decorators and MobX decorators in the same app. We recommend not to use MobX decorators for now in projects that use LitElement.
```

Author: Done.

For more information see link:https://mobx.js.org/enabling-decorators.html#enabling-decorators-[Enabling decorators] in MobX documentation.
====

== Creating a Data Store

Here is a minimal example of a MobX store that contains one link:https://mobx.js.org/observable-state.html[observable] property (`quote`) and one link:https://mobx.js.org/actions.html[action] (`setQuote()`) for updating that property. Here the store is initialized and exported on the last line so that you can import the same store instance into any view or component to access the shared state.

> .my-state.ts

Reviewer: ```suggestion
.`my-state.ts`
```

Author: Done.

----
import { makeAutoObservable } from 'mobx';

class MyState {
  quote: string = `Anything that can be derived from the application state,
    should be. Automatically. - MobX documentation`;

  constructor() {
    makeAutoObservable(this);
  }

  setQuote(quote: string) {
    this.quote = quote;
  }
}
export const myState = new MyState();
----

link:https://mobx.js.org/observable-state.html#makeautoobservable[makeAutoObservable] can automatically infer that `quote` is an observable and `setQuote()` is an action.

Make sure that an initial value is assigned to all state properties before calling `makeAutoObservable()`, otherwise the property will not be observable. This is a link:https://mobx.js.org/observable-state.html#limitations[MobX limitation] since we can't use the TypeScript configuration `"useDefineForClassFields": true` due to conflict with LitElement decorators mentioned above.

.What is a data store?
[NOTE]
====
"The main responsibility of stores is to move logic and state out of your components into a standalone testable unit." - link:https://mobx.js.org/defining-data-stores.html#stores[Defining data stores] in MobX documentation.
====

== Using a Store

> .my-view.ts

Reviewer: ```suggestion
.`my-view.ts`
```

Author: Done.

[source,typescript,subs="callouts+"]
----
import { customElement, html } from 'lit-element';
import { TextFieldElement } from '@vaadin/vaadin-text-field';
import { View } from '../view';
import { myState } from 'Frontend/store/my-state'; // <1>

@customElement('my-view')

> export class MyView extends View {

Reviewer 2: I think the `View` is from the starter wizard, not provided by Vaadin out-of-the-box. Maybe better to use `MobxLitElement` instead of `View`?

Author: Good point. I updated this to use `MobxLitElement` but mention in the callout text that you may instead use `View` if the project is generated from Start Vaadin.

  render() {
    const { quote } = myState; // <2>
    
    // <3>
    return html`
      <p>${quote}</p>
      <vaadin-text-field .value=${quote} @input=${this.onInput}></vaadin-text-field>
    `;
  }

  onInput(e: InputEvent) {
    myState.setQuote((e.target as TextFieldElement).value); // <4>
  }
}
----
<1> Import the store into your view or component.
<2> Alias one or more state properties from `myState` into local variables for easy referencing.
<3> Use state properties in the template.
<4> Update state by calling an action method in the store.

By default, it is not allowed to change the state outside of actions. Read more about actions in the MobX documentation: link:https://mobx.js.org/actions.html[Updating state using actions].

== Understanding MobX

> To be able to effectively use MobX and understand how it works, you should take a look at the official link:https://mobx.js.org/[MobX documentation] which is a great resource for learning the basic concepts as well as advanced usage.

Reviewer 3: Feels like a comma before which is necessary here:

```suggestion
To be able to effectively use MobX and understand how it works, you should take a look at the official link:https://mobx.js.org/[MobX documentation], which is a great resource for learning the basic concepts as well as advanced usage.
```

Author: Done. Thanks!
